### PR TITLE
Clean random generator

### PR DESCRIPTION
**Pull request by ternaus**
Fixes: https://github.com/albumentations-team/albumentations/issues/1874

## Summary by Sourcery

Refactor the random generator usage across the codebase to use numpy's default_rng, ensuring consistent and reproducible results. Enhance test coverage by adding new test cases for various augmentation functions, and fix issues related to random generator usage in tests.

Bug Fixes:
- Fix random generator usage to ensure consistent and reproducible results across various functions and tests.

Enhancements:
- Refactor random generator usage across the codebase to use numpy's default_rng for better performance and reproducibility.
- Improve test coverage by adding new test cases for various augmentation functions, ensuring they handle different scenarios and edge cases.

Tests:
- Add extensive test cases for cutout, filter_bboxes_by_holes, and other augmentation functions to validate their behavior with different parameters and inputs.
- Ensure all tests use numpy's default_rng for random number generation to maintain consistency in test results.

### CODE REVIEW

File: albumentations/core/transforms_interface.py [hunk 1]
```diff
@@ -85,6 +85,34 @@ def __init__(self, p: float = 0.5, always_apply: bool | None = None):
...
+        self.py_random = random.Random(self.seed)  # Create instance instead of using global
+
+    def set_random_state(self, seed: int | None) -> None:
+        """Set random state for this transform and all nested transforms.
+
+        Args:
+            seed: Random seed to use
+        """
+        if seed is not None:
+            random.seed(seed)  # Set standard library random seed
```
Comment (Erotemic) [inline albumentations/core/transforms_interface.py:99]: It looks like this implementation impacting the global random state? Is that necessary? I would have expected that calling the `set_random_state` on an object would only impact the object's internal state, and not the global state.

File: albumentations/core/composition.py [hunk 2]
```diff
@@ -120,6 +124,26 @@ def __init__(self, transforms: TransformsSeqType, p: float, mask_interpolation:
...
+        self.seed = seed
+        self.random_generator = np.random.default_rng(seed)
+
+        if seed is not None:
+            random.seed(seed)  # Set standard library random seed
+
+        # Propagate to all transforms
+        for transform in self.transforms:
+            if isinstance(transform, (BasicTransform, BaseCompose)):
+                transform.set_random_state(seed)
```
Comment (Erotemic) [inline albumentations/core/composition.py:146]: It looks like this is passing around the seed rather than the random generator. I think that might lead to undesirable behavior. For instance, say you compose two random affine transforms (a and b), and then you seed them with `seed=12345`. Each This would create two independent random states:

```
a = np.random.default_rng(12345)
b = np.random.default_rng(12345)
```

And then if you query them:

```
print(a.random())
print(b.random())
```

You will find they both print `0.22733602246716966`. Thus in the hypothetical case, the same affine transforms would be applied twice instead of having two independent affine transforms applied.

To fix this, I think it is sufficient to pass the `self.random_generator` instead of `seed`, e.g.:

```
    transform.set_random_state(self.random_generator)
```

This POC shows that it will cause the objects in the compose pipeline to share the same random state, which means the numbers they generate will be independent:

```
a = np.random.default_rng(12345)
b = np.random.default_rng(a)
print(a.random())
print(b.random())
```

Which produces:

```
0.22733602246716966
0.31675833970975287
```

Comment (ternaus) [reply albumentations/core/composition.py:146]: Good point, thanks. Will fix.

File: albumentations/core/transforms_interface.py (hunk 1 above)
Comment (ternaus) [reply albumentations/core/transforms_interface.py:99]: Cleaned this up in a later Pull request

File: albumentations/core/composition.py (hunk 2 above)
Comment (ternaus) [reply albumentations/core/composition.py:146]: Fixed in https://github.com/albumentations-team/albumentations/pull/2039
